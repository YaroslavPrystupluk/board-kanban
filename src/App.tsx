import { FC, KeyboardEvent, ChangeEvent, useState } from "react";
import { Spinner, Alert, ThemeProvider, Container } from "react-bootstrap";

import { useAppDispatch, useAppSelector } from "./hooks/hook";
import { fetchIssues } from "./store/slice/getIssuesSlice";
import SearchIssues from "./components/SearchIssues/SearchIssues";
import Breadcrumbs from "./components/Breadcrumbs/Breadcrumbs";
import Board from "./components/Board/Board";
import { DropResult } from "react-beautiful-dnd";

const App: FC = () => {
  const dispatch = useAppDispatch();
  const { issues, loading, error } = useAppSelector((state) => state.issues);

  const [repositoryUrl, setRepositoryUrl] = useState<string>("");

  const url = repositoryUrl.split("/");
  const owner = url[3];
  const repo = url[4];

  const handleKeyDoun = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key === "Enter") dispatch(fetchIssues([owner, repo]));
  };

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    setRepositoryUrl(e.target.value);
  };

  const showIssues = () => {
    dispatch(fetchIssues([owner, repo]));
  };

  const handleDragEnd = (result: DropResult) => {
    console.log("hello", result);
    const { source, destination, draggableId } = result;

    if (!destination) return;

    if (
      source.droppableId === destination.droppableId &&
      source.index === destination.index
    )
      return;

    // if()
  };

  return (
    <ThemeProvider
      breakpoints={["xxxl", "xxl", "xl", "lg", "md", "sm", "xs", "xxs"]}
      minBreakpoint="xxs"
    >
      <Container>
        <SearchIssues
          handleKeyDoun={handleKeyDoun}
          showIssues={showIssues}
          handleChange={handleChange}
          repositoryUrl={repositoryUrl}
        />
        {issues.length > 0 && <Breadcrumbs owner={owner} repo={repo} />}

        {loading ? (
          <Spinner
            style={{ position: "absolute", top: "50%", right: "50%" }}
            animation="border"
            variant="primary"
          />
        ) : error ? (
          <Alert style={{ textAlign: "center" }} variant="danger">
            Error: {error}
          </Alert>
        ) : (
          <Board handleDragEnd={handleDragEnd} />
        )}
      </Container>
    </ThemeProvider>
  );
};

export default App;
