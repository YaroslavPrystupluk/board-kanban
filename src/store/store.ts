import { configureStore } from "@reduxjs/toolkit";
import searchIssuesReducer from "./slice/getIssuesSlice";
import starsReducers from "./slice/getStarsSlice";

const store = configureStore({
	reducer: {
		issues: searchIssuesReducer,
		stars: starsReducers,
	},
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
