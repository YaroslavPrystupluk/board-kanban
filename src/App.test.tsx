import { describe, it, expect, vi, Mock } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider, useDispatch } from "react-redux";
import * as reduxHooks from "react-redux";
import App from "./App";
import store from "./store/store";

const mockState = {
	issues: {
		issues: [
			{ id: 1, state: "open", title: "Issue 1", user: { type: "user" } },
			{ id: 2, state: "in progress", title: "Issue 2", user: { type: "user" } },
			{ id: 3, state: "closed", title: "Issue 3", user: { type: "user" } },
		],
		loading: false,
		error: null,
	},
	stars: {
		stars: {
			stargazers_count: 100,
		},
	},
};

vi.mock("react-redux");
const mockedUseSelektor = vi.spyOn(reduxHooks, "useSelector");

describe("testing component 'App'", () => {
	it("renders app with issues and stars", async () => {
		mockedUseSelektor.mockReturnValue(mockState.issues.issues);
		const { container } = render(
			<Provider store={store}>
				<App />
			</Provider>,
		);

		expect(container).toMatchSnapshot();

		// expect(screen.queryByTestId("problem-list")).toBeNull();

		// const button = screen.getByRole("button", { name: /Load issues/i });
		// userEvent.click(button);
		// expect(dispatchMock).toHaveBeenCalledTimes(1);

		// await waitFor(() => {
		// 	expect(screen.getByTestId("problem-list")).toBeInTheDocument();
		// });

		// expect(screen.getByText("Issue 1")).toBeInTheDocument();
		// expect(screen.getByText("Issue 2")).toBeInTheDocument();
		// expect(screen.getByText("Issue 3")).toBeInTheDocument();
		// expect(screen.getByText("Stars: 100")).toBeInTheDocument();
	});

	// it("displays loading spinner", () => {
	// 	const loadingState = {
	// 		...mockState,
	// 		issues: {
	// 			...mockState.issues,
	// 			loading: true,
	// 		},
	// 	};

	// 	render(
	// 		<Provider store={store}>
	// 			<App />
	// 		</Provider>,
	// 	);

	// 	expect(screen.getByRole("status")).toBeInTheDocument();
	// });
});
