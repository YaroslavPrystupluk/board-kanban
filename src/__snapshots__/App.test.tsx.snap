// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`testing component 'App' > renders app with issues and stars 1`] = `<div />`;
